### PR TITLE
Kms upgrade

### PR DESCRIPTION
**Pull request by gannan08**

### CODE REVIEW

**Review by dmitrizagidulin — APPROVED** [review]
:+1:

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,24 @@
...
+  `bedrock-app-identity`.
+
+### Changed
+- **BREAKING**: Configuration of the kms is now under a `kms` namespace in the
+  config. The `config.tokenizer.kmsModule` configuration is now
+  `config.tokenizer.kms.defaultKmsModule`. The `config.tokenizer.kmsBaseUrl`
+  configuration is now `config.tokenizer.kms.baseUrl`.
+- **BREAKING**: Use the application's identity use `bedrock-app-identity` to
+  generate the `capabilityAgent` instead of a random seed during tokenizer
+  creation.
```
Comment (dlongley) [inline CHANGELOG.md]: Needs some rewording here:

```suggestion
- **BREAKING**: Use the application's identity use `bedrock-app-identity` to
  generate the `capabilityAgent` instead of a random seed during tokenizer
  creation.
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,5 +1,24 @@
...
+- **BREAKING**: Use random base58 encoded value for the `id` of the tokenizer
+  instead of the `id` from a randomly generated `capabilityAgent`.
```
Comment (dlongley) [inline CHANGELOG.md]: Why this change? IIRC, we want the tokenizer to have a did:key ID which is what the capability agent generated before.

File: lib/kms.js [hunk 3]
```diff
@@ -1,70 +1,54 @@
...
-    config.referenceId = referenceId;
+    keystoreConfig.referenceId = referenceId;
   }
   const {httpsAgent} = brHttpsAgent;
   return KmsClient.createKeystore({
-    url: `${kmsBaseUrl}/keystores`,
-    config,
+    url: `${bedrock.config.tokenizer.kms.baseUrl}/keystores`,
+    config: keystoreConfig,
+    invocationSigner: meterCapabilityInvocationSigner,
```
Comment (dlongley) [inline lib/kms.js]: This doesn't seem right, the invocation signer is for signing the zcap invocation for creating the keystore, it isn't for signing the meter zcap invocation (the KMS system will be signing that).

**Review by dlongley — CHANGES_REQUESTED** [review]

File: CHANGELOG.md (hunk 2 above)
Comment (gannan08) [reply CHANGELOG.md]: The previous behavior generated a random seed, then a capability agent from that seed.  This allowed the ID of the capability to change when creating a new tokenizer.  Now the capability agent ID comes from the application's identity. We can't use the capability agent ID because it no longer changes. What value should we use instead to ensure we uniquely identify new tokenizers?

File: CHANGELOG.md (hunk 1 above)
Comment (gannan08) [reply CHANGELOG.md]: The wording in the suggestion seems to be the same.

File: lib/kms.js (hunk 3 above)
Comment (gannan08) [reply lib/kms.js]: Will change the name to `invocationSigner`.

File: lib/tokenizers.js [hunk 4]
```diff
@@ -185,34 +184,39 @@ async function _readCurrentTokenizerRecord() {
...
 
-  // 1. Generate capability agent from handle and secret.
+  // 1. Generate capability agent from handle and application identity.
   const handle = 'primary';
-  const secret = base64url.decode(tokenizer.secret);
-  const capabilityAgent = await CapabilityAgent.fromSecret({handle, secret});
+  const {keys: {capabilityInvocationKey}} = getAppIdentity();
+  const keyPair = capabilityInvocationKey;
+  const signer = capabilityInvocationKey.signer();
+  const capabilityAgent = new CapabilityAgent({handle, signer, keyPair});
```
Comment (dlongley) [inline lib/tokenizers.js]: Hmm, unless I'm misunderstanding this, this seems like it makes it so that every tokenizer is the same now vs. having N tokenizers per app. We should discuss sometime tomorrow.

**Review by dlongley — COMMENTED** [review]
We should also remove promisify usage on bedrock-database calls while we're here.

File: lib/config.js [hunk 5]
```diff
@@ -8,10 +8,23 @@ const c = bedrock.util.config.main;
...
+
+// meter usage id to be used with the KMS
+config.tokenizer.kms.meterId = '';
 
 // default WebKMS base URL -- modify as needed for production systems where the
 // expectation is that the WebKMS runs on a different system from the system
 // running `bedrock-tokenizer` (note that this doesn't necessarily mean that
 // the systems won't use the same external domain)
-cc('tokenizer.kmsBaseUrl', () => `${config.server.baseUri}/kms`);
+cc('tokenizer.kms.baseUrl', () => `${config.server.baseUri}/kms`);
```
Comment (dlongley) [inline lib/config.js:30]: We need to add `tokenizer.kms` to the bedrock config ensure overrides section so that deployments are forced to specify all of its values.

File: lib/tokenizers.js [hunk 6]
```diff
@@ -235,12 +235,13 @@ async function _createTokenizer() {
...
-    {capabilityAgent, referenceId: 'primary'});
+  const keystore = await kms.createKeystore({
+    controller: capabilityAgent.id, referenceId: 'primary'
+  });
   const keystoreAgent = kms.getKeystoreAgent(
-    {capabilityAgent, keystore});
+    {capabilityAgent, keystoreId: keystore.id});
   const key = await keystoreAgent.generateKey(
-    {type: 'hmac', kmsModule: bedrock.config.tokenizer.kmsModule});
+    {type: 'hmac', kmsModule: bedrock.config.tokenizer.kms.defaultKmsModule});
```
Comment (dlongley) [inline lib/tokenizers.js]: `kmsModule` is no longer an option for `generateKey`, so this can be improved to a single line that just passes `{type: 'hmac'}` (github wouldn't let me suggest this).

File: test/mocha/mock.data.js [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+  // Use default webkms dev `id` and `serviceId`
+  id: 'urn:uuid:80a82316-e8c2-11eb-9570-10bf48838a29',
+  name: 'Example KMS',
+  service: {
+    // default dev `id` configured in `bedrock-kms-http`
+    id: 'did:key:z6MkwZ7AXrDpuVi5duY2qvVSx1tBkGmVnmRjDvvwzoVnAzC4',
+    type: 'webkms',
+  }
+}, {
+  // Use default edv dev `id` and `serviceId`
```
Comment (dlongley) [inline test/mocha/mock.data.js]: I don't think EDVs are used here -- do we still want this mock data since they aren't?

File: test/package.json [hunk 8]
```diff
@@ -4,36 +4,45 @@
...
   },
   "author": "",
   "dependencies": {
-    "bedrock": "^3.1.1",
-    "bedrock-account": "^3.0.0",
-    "bedrock-did-context": "^1.0.0",
-    "bedrock-express": "^3.2.0",
-    "bedrock-https-agent": "^2.0.0",
+    "bedrock": "^4.4.0",
+    "bedrock-account": "^5.0.0",
```
Comment (dlongley) [inline test/package.json]: Are bedrock-account, bedrock-passport, and bedrock-permission needed in the tests? If not, we should pare this down to what's needed.

**Review by dlongley — APPROVED** [review]
Thanks! Approving, but there are a still a few comments to address.

File: lib/config.js (hunk 5 above)
Comment (gannan08) [reply lib/config.js:30]: Acknowledged.

File: lib/tokenizers.js (hunk 6 above)
Comment (gannan08) [reply lib/tokenizers.js]: Got it.

File: test/mocha/mock.data.js (hunk 7 above)
Comment (gannan08) [reply test/mocha/mock.data.js]: Will remove.

File: test/package.json (hunk 8 above)
Comment (gannan08) [reply test/package.json]: Will pare down.

**Review by dlongley — APPROVED** [review]
